server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotsServiceIT.java

```diff
@@ -111,6 +112,42 @@ public void testSnapshotDeletionFailureShouldBeLogged() throws Exception {
         }
     }
 
+    public void testDeleteSnapshotWhenNotWaitingForCompletion() throws Exception {
+        createIndexWithRandomDocs("test-index", randomIntBetween(1, 5));
+        createRepository("test-repo", "mock");
+        createSnapshot("test-repo", "test-snapshot", List.of("test-index"));
+        MockRepository repository = getRepositoryOnMaster("test-repo");
+        PlainActionFuture<AcknowledgedResponse> listener = new PlainActionFuture<>();
+        repository.blockOnDataFiles();
+        try {
+            clusterAdmin().prepareDeleteSnapshot("test-repo", "test-snapshot").setWaitForCompletion(false).execute(listener);
+            listener.get(5, TimeUnit.SECONDS);
+            assertNotNull(getSnapshot("test-repo", "test-snapshot"));
+        } finally {
+            repository.unblock();
+        }
+        assertBusy(() -> assertThrows(SnapshotMissingException.class, () -> getSnapshot("test-repo", "test-snapshot")));
```

Comment:
I'd rather we waited for the deletion to complete by watching the cluster state on the master instead of just busy-waiting like this.